src/main/webapp/app/exercises/shared/structured-grading-criterion/grading-instructions-details/grading-instructions-details.component.ts

```diff
@@ -285,4 +320,113 @@ export class GradingInstructionsDetailsComponent implements OnInit {
+        const backupInstructionIndex = this.backupExercise.gradingCriteria![criteriaIndex].structuredGradingInstructions.findIndex((sgi) => {
+            return sgi.id === instruction.id;
+        });
+
+        this.exercise.gradingCriteria![criteriaIndex].structuredGradingInstructions![instructionIndex] = cloneDeep(
+            this.backupExercise.gradingCriteria![backupCriteriaIndex].structuredGradingInstructions![backupInstructionIndex],
+        );
+    }
+
+    /**
+     * @function deleteInstruction
+     * @desc Deletes selected instruction
+     * @param instruction {GradingInstruction} the instruction which should be deleted
+     * @param criteria {GradingCriterion} the criteria, which includes the instruction that will be deleted
+     */
+    deleteInstruction(instruction: GradingInstruction, criteria: GradingCriterion) {
+        const criteriaIndex = this.exercise.gradingCriteria!.indexOf(criteria);
+        const instructionIndex = this.exercise.gradingCriteria![criteriaIndex].structuredGradingInstructions.indexOf(instruction);
+        this.exercise.gradingCriteria![criteriaIndex].structuredGradingInstructions.splice(instructionIndex, 1);
+    }
+
+    /**
+     * @function addNewInstruction
+     * @desc Adds new grading instruction for desired grading criteria
+     * @param criteria {GradingCriterion} the criteria, which includes the instruction that will be inserted
+     */
+    addNewInstruction(criteria: GradingCriterion) {
+        const criteriaIndex = this.exercise.gradingCriteria!.indexOf(criteria);
+        const instruction = new GradingInstruction();
+        this.exercise.gradingCriteria![criteriaIndex].structuredGradingInstructions.push(instruction);
+    }
+
+    /**
+     * @function addNewGradingCriteria
+     * @desc Adds new grading criteria for the exercise
+     */
+    addNewGradingCriteria() {
+        const criteria = new GradingCriterion();
+        criteria.structuredGradingInstructions = [];
+        criteria.structuredGradingInstructions.push(new GradingInstruction());
+        this.exercise.gradingCriteria!.push(criteria);
+    }
+
+    /**
+     * @function onCriteriaTitleChange
+     * @desc Detects changes for grading criteria title
+     * @param {GradingCriterion} criteria the criteria, which includes title that will be changed
+     */
+    onCriteriaTitleChange($event: any, criteria: GradingCriterion) {
+        const criteriaIndex = this.exercise.gradingCriteria!.indexOf(criteria);
+        this.exercise.gradingCriteria![criteriaIndex].title = $event.target.value;
+    }
+
+    /**
+     * @function resetCriteriaTitle
+     * @desc Resets the whole grading criteria title
+     * @param criteria {GradingCriterion} the criteria, which includes title that will be reset
+     */
+    resetCriteriaTitle(criteria: GradingCriterion) {
+        const criteriaIndex = this.exercise.gradingCriteria!.findIndex((gradingCriteria) => {
+            return gradingCriteria.id === criteria.id;
+        });
```

Comment:
You use index finding a lot for these grading criteria so it can be turned into a function